Fix for ambiguous tupler

blast-hardcheese opened the pull request:
Resolves #713

Initially I thought it was as simple as bumping `TuplerAppend` to `[4..21`, but I was apparently mistaken. I don't _quite_ understand why 3 is the magic number here, but I'm perfectly happy to revise my approach to keep `1And4` if there's a preferred way to approach this.

Currently I'm stuck, as the instances are defined on `Tupler`'s companion, so I can't even disable the offending implicit via name collision.

(PS: I had to recreate this PR entirely, as apparently force pushing is disallowed on this repo -- I've never seen that setting used before, could someone kindly explain why it is desirable?)

julienrf commented:
Thank you @blast-hardcheese, I will have a look this week.

blast-hardcheese commented:
For future contributors, the one-liner for generating the unit tests here is:

```scala
@ val terms = List("a", "b", "c", "d", "e", "f", "g", "h", "i").map(Term.Name(_))
terms: List[Term.Name] = List(Term.Name("a"), Term.Name("b"), Term.Name("c"), Term.Name("d"), Term.Name("e"), Term.Name("f"), Term.Name("g"), Term.Name("h"), Term.Name("i"))

@
println(q"""..${terms.reverse.tails.map(_.reverse).flatMap(xs => (1 until xs.length).map(xs.splitAt(_)).map({ case (fst, snd) => q"""tupling(${if (fst.length > 1) { q"(..${fst})" } else { fst.head } }, ${if (snd.length > 1) { q"(..${snd})" } else { snd.head } }): (..${xs.map { case Term.Name(s) => Type.Name(s.toUpperCase) } })""" })).toList}""".syntax)
{
  tupling(a, (b, c, d, e, f, g, h, i)): (A, B, C, D, E, F, G, H, I)
  tupling((a, b), (c, d, e, f, g, h, i)): (A, B, C, D, E, F, G, H, I)
  tupling((a, b, c), (d, e, f, g, h, i)): (A, B, C, D, E, F, G, H, I)
  ...
```

then resolving the compiler errors (as the return type ascriptions are all the same, most of them will be wrong, but the compiler will tell you what the correct ones are). This was just to cover enough area to try to reduce the likelihood of this in the future.

julienrf commented on json-schema/json-schema/src/test/scala/endpoints4s/TuplerTests.scala:
```diff
@@ -5,18 +5,45 @@ class TuplerTests {
-    tupling((a, b, c, d, e), f): (A, B, C, D, E, F)
-    tupling(a, ()): A
-    tupling((), a): A
+    tupling(a, (b, c, d, e))  : (A, (B, C, D, E))
```
We should get a flat tuple of type `(A, B, C, D, E)` here. The line you deleted from `Tupler.scala` should produce such a result.

blast-hardcheese replied:
Yep, confirmed. I'll not break this with further work.

julienrf commented on json-schema/json-schema/src/test/scala/endpoints4s/TuplerTests.scala:
```diff
@@ -5,18 +5,45 @@ class TuplerTests {
-    tupling((), a): A
+    tupling(a, (b, c, d, e))  : (A, (B, C, D, E))
+    tupling((a, b, c, d, e), f)  : (A, B, C, D, E, F)
+    tupling((a, b, c, d), (e, f)): (A, B, C, D, (E, F))
```
Here as well, I would like to return a flat tuple `(A, B, C, D, E, F)`

blast-hardcheese replied:
This would be a change to tupler. I've updated the PR with a reduced diff in favor of easy reviewing, I'll re-sort towards the end of this process.

If at all possible, it would be ideal to permit rebasing and force-push for branches, but if you'll squash at the end I'll just not be concerned about producing unnecessary intermediate commits here.

blast-hardcheese replied:
Also, sorry, to what end are we doing this?

You asked about
```scala
    tupling((a, b, c, d), (e, f)): (A, B, C, D, (E, F))
```
but
```scala
    tupling((a, b, c), (d, e, f)): (A, B, C, (D, E, F))
    tupling((a, b), (c, d, e, f)): ((A, B), C, D, E, F)
    tupling(a, (b, c, d, e, f)): (A, (B, C, D, E, F))
```
would also not be covered by this. Isn't this just that we only have so many tupler extensions so far? [I've worked around this when attempting to generate tuple structures compatible with the current Tuple semantics](https://github.com/twilio/guardrail/blob/v0.62.0/modules/codegen/src/main/scala/com/twilio/guardrail/generators/Scala/EndpointsClientGenerator.scala#L290-L294), so I'd err on the side of stability until we have the truly generic solution, if I'm being selfish.

The change in order to make this test pass is
```diff
diff --git a/json-schema/json-schema/src/main/scala/endpoints4s/Tupler.scala b/json-schema/json-schema/src/main/scala/endpoints4s/Tupler.scala
index 929d19e8..88bfcfb2 100644
--- a/json-schema/json-schema/src/main/scala/endpoints4s/Tupler.scala
+++ b/json-schema/json-schema/src/main/scala/endpoints4s/Tupler.scala
@@ -114,7 +114,23 @@ trait Tupler3 extends Tupler2 {
 
 }
 
-trait TuplerLeftUnit extends TuplerAppend {
+trait Tupler4 extends TuplerAppend {
+
+  implicit def tupler4And2[A, B, C, D, E, F]
+      : Tupler[(A, B, C, D), (E, F)] { type Out = (A, B, C, D, E, F) } =
+    new Tupler[(A, B, C, D), (E, F)] {
+      type Out = (A, B, C, D, E, F)
+      def apply(abcd: (A, B, C, D), ef: (E, F)): (A, B, C, D, E, F) =
+        (abcd._1, abcd._2, abcd._3, abcd._4, ef._1, ef._2)
+      def unapply(out: (A, B, C, D, E, F)): ((A, B, C, D), (E, F)) = {
+        val (a, b, c, d, e, f) = out
+        ((a, b, c, d), (e, f))
+      }
+    }
+
+}
+
+trait TuplerLeftUnit extends Tupler4 {
 
   implicit def leftUnit[A]: Aux[Unit, A, A] =
     new Tupler[Unit, A] {
diff --git a/json-schema/json-schema/src/test/scala/endpoints4s/TuplerTests.scala b/json-schema/json-schema/src/test/scala/endpoints4s/TuplerTests.scala
index 67c00c4d..4f630166 100644
--- a/json-schema/json-schema/src/test/scala/endpoints4s/TuplerTests.scala
+++ b/json-schema/json-schema/src/test/scala/endpoints4s/TuplerTests.scala
@@ -20,7 +20,7 @@ class TuplerTests {
 
     tupling(a, (b, c, d)): (A, (B, C, D))
     tupling((a, b, c), (d, e)): (A, B, C, (D, E))
-    tupling((a, b, c, d), (e, f)): (A, B, C, D, (E, F))
+    tupling((a, b, c, d), (e, f)): (A, B, C, D, E, F)
     tupling((a, b, c), (d, e, f)): (A, B, C, (D, E, F))
     tupling((a, b), (c, d, e, f)): ((A, B), C, D, E, F)
     tupling(a, (b, c, d, e, f)): (A, (B, C, D, E, F))
```
but adding the various other combinations, `1And5`, `3And3`, `4And2`, made a bunch of other tests fail so I did not include it here.

julienrf replied:
> Also, sorry, to what end are we doing this?
> 
> You asked about
> 
> ```scala
>     tupling((a, b, c, d), (e, f)): (A, B, C, D, (E, F))
> ```
> 
> but
> 
> ```scala
>     tupling((a, b, c), (d, e, f)): (A, B, C, (D, E, F))
>     tupling((a, b), (c, d, e, f)): ((A, B), C, D, E, F)
>     tupling(a, (b, c, d, e, f)): (A, (B, C, D, E, F))
> ```
> 
> would also not be covered by this. Isn't this just that we only have so many tupler extensions so far?

The goal of `Tupler` is to concatenate tuples, so that we end-up with a flat tuple instead of nested tuples (and occurrences of `Unit` are removed). The cases you mention should be fixed as well… The problem is that the `Tupler` instances currently cover only the most common cases, and being exhaustive would require a lot of implicit definitions (and headaches).

I suggest that we remove these lines from the test, as this is not something we want to support (I believe).

blast-hardcheese replied:
Maybe split them out into a "stability" section? The reason for this PR in the first place is because an otherwise helpful addition ended up making some valid usecases completely unrepresentable.

At some point this breaks down, for instance a `Tuple22` combined with anything. I've needed to work around this for particularly complex OpenAPI specifications, which is what led me to `endpoints` in the first place.

Either way, your points about defining disambiguating `Tupler` instances in my own project are completely valid, just looking to wrap up this issue to help out others that may run into this same problem before the HList matter is settled.

I'll move the tests to a separate section that can act as a canary for when things change, but if you feel they should be removed I'm perfectly happy to do that as well.

Thanks again for your time here.

julienrf reviewed (commented):
Thanks @blast-hardcheese for investigating this issue.

I don’t think it is safe or correct to remove `tupler1and4`. Unfortunately, the current approach based on tuples and lots of implicit definitions does not scale.

I guess the “best” way to solve your problem would be to add more specific implicit definitions for `Tupler[(A, B, C), (D, E, F, G)]` and so on for your use cases… You can do that in your project, by defining these implicit definitions in a trait and mixing this trait in your algebra.

Ultimately, I believe the right way to address the issue would be to switch to an inductive representation of tuples---an `HList`, so that we could just concatenate them with a few rules. This idea is tracked by #60. I’m not sure whether we should wait one more year before we can leverage Scala 3 tuples, or if we should fix the problem now by using shapeless.

blast-hardcheese commented:
Understood, regarding 1and4.

To justify my attempt to remove `1and4`, it seemed like that was the most straightforward way to at least make all representations compile out of the box, as there are legitimate programs that are no longer representable past 0.9.

I'll see if I can resolve this by addressing your feedback here, as well as fitting new tuplers in the hierarchy in order to resolve the ambiguity.

blast-hardcheese commented:
@julienrf Anyway, ready for re-review. If you'd like more changes, I've got a better understanding of why the previous mechanics were not behaving as expected. I can also reorder the tests to be coherent when you feel comfortable with where this PR stands.

Thanks

julienrf commented on json-schema/json-schema/src/test/scala/endpoints4s/TuplerTests.scala:
```diff
@@ -5,19 +5,56 @@ class TuplerTests {
+    tupling((a, b, c, d, e, f, g), h): (A, B, C, D, E, F, G, H)
+    tupling((a, b, c, d, e, f, g, h), i): (A, B, C, D, E, F, G, H, I)
+
+    // Stability tests
```
Not sure what you mean by “stability” test, do you mind elaborating?

blast-hardcheese replied:
"Stability of implementation", whether a proposed change alters the way things stand today. Any alterations to these "out of spec" tests would be reflected in a change to these lines, signaling to both maintainers and users of this library when the semantics change.

julienrf commented:
Thanks for your help and patience @blast-hardcheese!
